fivox/spikeLoader.cpp

```diff
@@ -60,62 +59,40 @@ class SpikeLoader::Impl
+        _spikesEnd = _report->getEndTime();
+        // We need _hasEnded to return true  is file based reports because
+        // otherwise getFrameRange doesn't work as expected before the end
+        // of the report is reached. The only way to detect this with the new
+        // API is by checking if the endTime is already different from 0 at
+        // construction, because for streams it will always be 0. This is
+        // relying on an implementation detail of the spikeReport, so it's
+        // worth considering to include this detail as part of the API or add
+        // the isStream function again;
+        if( _spikesEnd != 0 )
+            _isFile = true;
     }
 
     void updateTimeRange()
     {
         // Streaming in progress; Only report fully finished frames.
-        if( _spikesReader && !_spikesReader->hasEnded( ))
-        {
-            lunchbox::ScopedWrite mutex( _getSpikesLock );
-            const brain::Spikes& spikes = _spikesReader->getSpikes();
-            if( !spikes.empty( ))
-                // don't update _spikesStart to calculate absolute frame numbers
-                // see https://bbpcode.epfl.ch/code/#/c/19337
-                _spikesEnd = spikes.getEndTime();
-        }
-    }
-
-    void _loadSpikes( const URI& spikes )
-    {
-        if( _loadBinarySpikes( spikes.getPath( )))
+        if( !_report )
             return;
 
-        LBINFO << "No valid binary .spikes file found, loading from .dat..."
-               << std::endl;
-        _spikesReader.reset( new brain::SpikeReportReader( spikes ));
-        _spikesStart = _spikesReader->isStream() ? 0.f :
-                                                _spikesReader->getStartTime();
-        _spikesEnd = _spikesReader->isStream() ? 0.f :
-                                                _spikesReader->getEndTime();
-    }
-
-    bool _loadBinarySpikes( const std::string& spikes )
-    {
-        std::unique_ptr< lunchbox::MemoryMap > spikesFile
-                ( new lunchbox::MemoryMap( spikes ));
-        const size_t size = spikesFile->getSize();
-        if(( size % sizeof( uint32_t )) != 0 )
-            return false;
-        const size_t nElems = size / sizeof( uint32_t );
-        const uint32_t* iData = spikesFile->getAddress< uint32_t >();
-        size_t index = 0;
-
-        const uint32_t magic = 0xf0a;
-        const uint32_t version = 1;
-        if( index >= nElems || iData[ index++ ] != magic )
-            return false;
-        if( index >= nElems || iData[ index++ ] != version )
-            return false;
-
-        _spikesFile = std::move( spikesFile );
-        const float* fData = _spikesFile->getAddress< float >();
-        _spikesStart = fData[index]; // first spike timestamp after header
-        _spikesEnd = fData[nElems - 2]; // last spike timestamp
-        return true;
+        const auto endtime = _report->getEndTime( );
+        lunchbox::ScopedWrite mutex( _getSpikesLock );
```

Comment:
This should be fixed in the datasource imo. This one calls update() (which calls for the frame/time range) and samples() (which triggers load) concurrently. It has to be protected there, not in the loader.